### PR TITLE
add lfzip-nlms and libbsc recipe

### PR DESCRIPTION
**Pull request by shubhamchandak94**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml"
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by shubhamchandak94** [comment]
@conda-forge-admin, please ping team @conda-forge/help-python-c for pull request at https://github.com/conda-forge/staged-recipes/pull/11190

File: recipes/lfzip-nlms/build.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+mkdir -p $PREFIX/bin
+cd src/
+cd libbsc/
+make CC=$CXX
+cd ../
+$CXX nlms_helper.cpp -std=c++11 -o nlms_helper.out -Wall -O3 -march=native
```
Comment (isuruf) [inline recipes/lfzip-nlms/build.sh:8]:
```suggestion
$CXX nlms_helper.cpp -std=c++11 -Wall -O3 $CFLAGS -o nlms_helper.out $LDFLAGS
```

File: recipes/lfzip-nlms/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}
+    - llvm-openmp  # [osx]
+    - libgomp      # [linux]
+    - make
+  host:
+    - python
+    - pip
```
Comment (isuruf) [inline recipes/lfzip-nlms/meta.yaml:26]: You don't need `python` and `pip` to build the software right?

Comment (shubhamchandak94) [reply recipes/lfzip-nlms/meta.yaml:26]: Not sure, I just need numpy during execution, so I thought I needed to put it here.

**Comment by isuruf** [comment]
>  Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)

This box is ticked, but `lfzip-nlms` is vendoring `libbsc` right?

File: recipes/lfzip-nlms/meta.yaml (hunk 2 above)
Comment (isuruf) [reply recipes/lfzip-nlms/meta.yaml:26]: If so, you need it only in `run`. You can keep `host` empty.

File: recipes/lfzip-nlms/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    - python
+    - numpy
+
+test:
+  source_files:
+      - data/evaluation_datasets/gas/HT_Sensor_dataset_combined.npy
+
+about:
+  home: https://github.com/shubhamchandak94/LFZip
+  license: MIT
```
[the remark this answers is not in the document]
Comment (shubhamchandak94) [reply recipes/lfzip-nlms/meta.yaml:37]: I wanted to put MIT license on my code. Does including BSC (which is not my code) force me to use Apache-2.0? I don't have a strong preference

[the remark this answers is not in the document]
Comment (isuruf) [reply recipes/lfzip-nlms/meta.yaml:37]:
```suggestion
  license: MIT and Apache-2.0
```

File: recipes/lfzip-nlms/meta.yaml (hunk 2 above)
Comment (shubhamchandak94) [reply recipes/lfzip-nlms/meta.yaml]: Thanks, I have removed the host section and the tests still succeed.

File: recipes/lfzip-nlms/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  skip: True  # [win]
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}
+    - llvm-openmp  # [osx]
+    - libgomp      # [linux]
+    - make
+  run:
+    - python
```
Comment (isuruf) [inline recipes/lfzip-nlms/meta.yaml:25]:
```suggestion
    - python >=3
```

Comment (shubhamchandak94) [reply recipes/lfzip-nlms/meta.yaml]: I got a linting error and am not sure how to correct it. Should I add noarch?
Non noarch packages should have python requirement without any version constraints.

Comment (isuruf) [reply recipes/lfzip-nlms/meta.yaml]: You can ignore the linter error. It's buggy.

Comment (shubhamchandak94) [reply recipes/lfzip-nlms/meta.yaml]: Thanks!

**Comment by isuruf** [comment]
Can you create a new recipe for `libbsc`?

**Comment by shubhamchandak94** [comment]
> Can you create a new recipe for `libbsc`?

Is that allowed if I am not the author of libbsc?

**Comment by isuruf** [comment]
> Is that allowed if I am not the author of libbsc?

Yes. Anybody can be a maintainer.

**Comment by shubhamchandak94** [comment]
Thanks. Am I also allowed to still keep the libbsc code on the github repository? I want that for people who want to directly work with the source code and use other modes in LFZip that are not part of the conda installation.

**Comment by isuruf** [comment]
> Am I also allowed to still keep the libbsc code on the github repository?

That's fine.

> Also, I am using libbsc as an executable rather than importing it as a library (cp libbsc/bsc $PREFIX/bin/libbsc in my build.sh). Will that still work if I make libbsc a separate package?

Yes, `lfzip-nlms` will depend on `libbsc` package.

> The pull request template says "If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged". Since I have included the libbsc license, it might be simpler to just keep it as it is instead of creating a new conda package.

I strongly recommend different conda packages for different packages. Vendoring is okay if the vendored package needs significant changes, but it looks like you are using the upstream source and build system.

**Comment by shubhamchandak94** [comment]
Got it. Thanks. I will make a new PR for libbsc and point you to that for review

**Comment by isuruf** [comment]
You can add the recipe to this PR too if you like

File: recipes/libbsc/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+{% set name = "libbsc" %}
+{% set version = "3.1.0a" %}
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: https://github.com/shubhamchandak94/libbsc/archive/v{{ version }}.tar.gz
```
Comment (isuruf) [inline recipes/libbsc/meta.yaml:11]: Can you use upstream source instead of your fork?

Comment (shubhamchandak94) [reply recipes/libbsc/meta.yaml:11]: I actually made a few minor fixes related to multithreading in my fork, that's why I've been sticking to it. I am happy to provide appropriate credit to the original author if you think that's needed.

Comment (isuruf) [reply recipes/libbsc/meta.yaml:11]: Can you use upstream source and add your changes as a patch file so that anybody can see what changed?

Comment (shubhamchandak94) [reply recipes/libbsc/meta.yaml:11]: I can do that. Should the patch be put in the recipe folder?

Comment (isuruf) [reply recipes/libbsc/meta.yaml:11]: See https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#patches

Comment (shubhamchandak94) [reply recipes/libbsc/meta.yaml:11]: I am working on this and have one more question. The last release on the upstream github repo (https://github.com/IlyaGrebnov/libbsc) is quite old. There have been 8 reasonably large commits on the upstream repo and 3 on my fork after that. Is it okay to include all these in a patch file (size 686KB)?

Comment (shubhamchandak94) [reply recipes/libbsc/meta.yaml]: Pushed it now.

Comment (isuruf) [reply recipes/libbsc/meta.yaml]: Have you talked with upstream authors about making a release?

Comment (shubhamchandak94) [reply recipes/libbsc/meta.yaml]: I don't know the person and I know someone who contacted them (for some other reason) and didn't get a response. If you think that's required, I can create an issue there.

Comment (isuruf) [reply recipes/libbsc/meta.yaml]: Yes, please contact upstream about a release.

File: recipes/libbsc/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,43 @@
+# Note: there are many handy hints in comments in this example -- remove them when you've finalized your recipe
+
+{% set name = "libbsc" %}
+{% set version = "3.1.0" %}
```
Comment (isuruf) [inline recipes/libbsc/meta.yaml:4]:
```suggestion
{% set version = "3.1.1.a0" %}
```

File: recipes/libbsc/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: https://github.com/IlyaGrebnov/libbsc/archive/{{ version }}.tar.gz
```
Comment (isuruf) [inline recipes/libbsc/meta.yaml:11]: You can make the following change after changing the version as above.
```suggestion
  url: https://github.com/IlyaGrebnov/libbsc/archive/3dea3471251fbac72c895f42f22a1aa75b81da33.tar.gz
```

File: recipes/libbsc/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  sha256: 6645d95d929ffe9c5adcdbbdcd7151aa122356640ef4ece90d5b62e3d48fd007
+  patches:
+    - latest_github_version_plus_multithreading.patch
```
Comment (isuruf) [inline recipes/libbsc/meta.yaml:14]: After doing the change above, you can add your changes here.

Comment (isuruf) [reply recipes/libbsc/meta.yaml:14]: I meant only your changes.

Comment (shubhamchandak94) [reply recipes/libbsc/meta.yaml]: I made the suggested updates and pushed the code

File: recipes/libbsc/multithreading.patch [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
++        numThreads = paramnumthr;
+         if (numThreads <= nBlocks) paramEnableMultiThreading = 0;
+         if (numThreads >= nBlocks) numThreads = nBlocks;
+     }
+@@ -717,7 +718,7 @@ void ShowUsage(void)
+     fprintf(stdout, "  -P       Enable large 2MB RAM pages, default: disable\n");
+ #endif
+ #ifdef LIBBSC_OPENMP
+-    fprintf(stdout, "  -t       Disable parallel blocks processing, default: enable\n");
++    fprintf(stdout, "  -t<num_thr>       Number of threads: default 1\n");
```
Comment (isuruf) [inline recipes/libbsc/multithreading.patch:45]: This is a breaking change. Can you make the default to be `omp_get_max_threads()`?

Comment (shubhamchandak94) [reply recipes/libbsc/multithreading.patch]: I have made another commit removing the patch. I realized that I was not using the features of the patch in this project, so it's probably better to just go with the original.

File: recipes/lfzip-nlms/build.sh [hunk 10]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+mkdir -p $PREFIX/bin
+cd src/
+$CXX nlms_helper.cpp -std=c++11 -Wall -O3 $CFLAGS -o nlms_helper.out $LDFLAGS
+cp nlms_helper.out $PREFIX/bin
+(echo "#!/usr/bin/env python3" && cat nlms_compress.py) > nlms_compress_1.py
```
Comment (isuruf) [inline recipes/lfzip-nlms/build.sh:7]: Can this be done upstream?

Comment (shubhamchandak94) [reply recipes/lfzip-nlms/build.sh:7]: Sorry, I don't understand what upstream refers to here

File: recipes/lfzip-nlms/build.sh [hunk 11]
```diff
@@ -0,0 +1,10 @@
...
+# fix bsc path from dependency
+sed "s/BSC_PATH = .*/BSC_PATH = 'bsc'/g" nlms_compress_1.py > $PREFIX/bin/lfzip-nlms
```
Comment (isuruf) [inline recipes/lfzip-nlms/build.sh:9]: Can this be fixed upstream?
i.e. fall back to `bsc` from PATH if the file doesn't exist in the  usual location

File: recipes/lfzip-nlms/build.sh (hunk 10 above)
Comment (isuruf) [reply recipes/lfzip-nlms/build.sh:7]: I mean at https://github.com/shubhamchandak94/LFZip

File: recipes/lfzip-nlms/build.sh (hunk 11 above)
Comment (shubhamchandak94) [reply recipes/lfzip-nlms/build.sh]: Thanks, I've fixed that now.

File: recipes/libbsc/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    - make
+  host:
+  run:
+
+test:
+  commands:
+      - bsc -h
+
+about:
+  home: https://github.com/shubhamchandak94/libbsc
```
Comment (isuruf) [inline recipes/libbsc/meta.yaml:32]: Can you fix the link here?

Comment (shubhamchandak94) [reply recipes/libbsc/meta.yaml]: Done

**Pull request merged**